[Allard] ip

AllardQuek opened the pull request:
# Duke ~~University~~ Pro

<div align="center">
  <img src="https://cdn.wallpapersafari.com/90/82/IFSdjh.jpg" width=400 />
</div>

As [Duke Erwin](https://twitter.com/Duker_PDX) once said:

> Being busy,
> is not the same as *being productive*

## Some exciting features.
* Add ToDos
* Add Tasks with deadlines
* Add Events 

## How to use.
1. Download the jar file from releases.
2. Move the jar file into your desired directory.
3. On your command line, navigate to that directory and run `java -jar ip.jar`


### Code Snippet!
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
        System.out.println("Yay, added a fenced code block with syntax highlighting.")
    }
}
```

## ToDo for ToDo
- [X] Weeks 1 to 2
- [ ] Pull Request for Week 3
- [ ] Week 4
- [ ] Improvements to previous work

<br />
<div align="center">
<p> Please let me know any suggestions for improvement!</p>
🙏
</div>

Commit:
Add Gradle support

AllardQuek committed:
ignore class files and dump file

AllardQuek committed:
first attempt Level-1

AllardQuek committed:
add task component

AllardQuek committed:
Store text entered and display tasks when requested.

AllardQuek committed:
Complete Level 2

AllardQuek committed:
add done field for Task

AllardQuek committed:
Complete Level-3

Add the ability to mark tasks as done.

AllardQuek committed:
add task types

Todo, Deadline and Event

AllardQuek committed:
add simple input test case

AllardQuek committed:
Complete Level-4

Add support for tracking three types of tasks: ToDos, Deadlines, Events

AllardQuek committed:
add exception class

AllardQuek committed:
Complete Level-5

Teach Duke to deal with errors such as incorrect inputs entered by the user.

AllardQuek committed:
Complete Level-6: Delete

Add support for deleting tasks from the list.

AllardQuek committed:
Modify arraylist

AllardQuek committed:
Update tests

Included tests for all features and invalid inputs

AllardQuek committed:
Added documentation and reorganised code

AllardQuek committed:
Added greetUser method

AllardQuek committed:
Fix if statements styling

AllardQuek committed:
Add TaskList component

AllardQuek committed:
Add data file for tasks

AllardQuek committed:
Edit toString formatting

AllardQuek committed:
Add Storage class to save and load tasks

AllardQuek committed:
Change flow to include run method

AllardQuek committed:
Add Parser and Ui components to be completed

AllardQuek committed:
Add documentation for classes

AllardQuek committed:
Merge branch 'branch-level-7'

Complete Level-7: Save tasks to hard disk

AllardQuek committed:
Update with datetime format for deadline

AllardQuek committed:
Store deadline as LocalDate

AllardQuek committed:
Rename time field to dateTime

AllardQuek committed:
Factor out getTask and getDateTime

AllardQuek committed:
Ignore dump folder

AllardQuek committed:
Handle LocalDate deadlines

AllardQuek committed:
Merge branch 'branch-level-8'

Completed level-8: Dates and Times

AllardQuek committed:
Use more specific method names

AllardQuek committed:
Rename filepath to filePath

AllardQuek committed:
Add handleDelete and handleDone

AllardQuek committed:
Add Ui component to deal with user interactions

AllardQuek committed:
Add Parser component to understand user input

AllardQuek committed:
Add constructor for empty TaskList

AllardQuek committed:
Factor out parsing and ui functionality

AllardQuek committed:
Add documentation

AllardQuek committed:
Divide classes into packages

AllardQuek committed:
Package the App as a JAR file

AllardQuek committed:
Add tests for Deadline and ToDo

AllardQuek committed:
Delete gradle build

AllardQuek committed:
Add docs for classes and methods

AllardQuek committed:
Improve test name

AllardQuek committed:
Merge branch 'A-JavaDoc'

Add documentation to classes and methods.

AllardQuek committed:
Add some edits to follow coding standard

AllardQuek committed:
Complete Level-9: Find tasks

AllardQuek committed:
Merge branch 'branch-A-CodingStandard'

Edits to follow a coding standard.

AllardQuek committed:
Complete Level-9: Find tasks

AllardQuek committed:
Remove raw error output

AllardQuek committed:
Merge branch 'branch-Level-9'

Complete level 9: find tasks

AllardQuek committed:
Update case where no file to save to

AllardQuek commented:
# DukePro

AllardQuek committed:
Order imports

AllardQuek committed:
Merge remote-tracking branch 'origin/add-gradle-support'

Start adding gradle to project

AllardQuek committed:
Add test to check if Deadline is marked done correctly

AllardQuek committed:
Tweak build for project

AllardQuek committed:
Add checkstyle config

AllardQuek committed:
Enable checkstyle in Gradle

AllardQuek committed:
Ignore hidden files

AllardQuek committed:
Fix some checkstyle

AllardQuek committed:
Add sample user images

AllardQuek committed:
Add fxml files

AllardQuek committed:
Add fxml dependencies

AllardQuek committed:
Add fxml components

AllardQuek committed:
Set up fxml entry point

AllardQuek committed:
Remove unused import

AllardQuek committed:
Change return types and access modifiers

AllardQuek committed:
Change return type from boolean to string

AllardQuek committed:
Allow Duke to be run using text-based UI or GUI

AllardQuek committed:
Test with some user input

darren2pro commented on src/main/java/duke/DialogBox.java:
```diff
@@ -0,0 +1,61 @@
+import javafx.scene.layout.HBox;
+
+/**
+ * An example of a custom control using FXML.
```
Maybe you could consider removing this line because "An example of a custom control..." might be slightly misleading when this is your actual class.

AllardQuek replied:
True!

darren2pro commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,73 @@
+public class Duke {
+    public static Storage storage;
+    public static TaskList tasks;
+    private final String STORAGE_PATH = "data/duke.txt";
```
I like that you saved this as a constant.

darren2pro reviewed (commented):
Looks good overall! 😄

stephanie-csy commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,73 @@
+
+    /**
+     * Returns the response from Duke given the user input to be shown on the GUI.
+     */
```
Perhaps missing a description for param?

stephanie-csy reviewed (commented):
Looks good to me! Except:

1. Missing/incomplete javadocs for a few public methods
2. Basic indentation should be 4 spaces but some big portions of your code look like the indentation is 8 spaces, while the rest look like 4 spaces. 

But overall I like how neat and easily readable it is!

AllardQuek commented:
Thanks @stephanie-csy and @darren2pro! :)

AllardQuek committed:
Fix a few coding standards issues

AllardQuek committed:
Comment out some rules

AllardQuek committed:
Enable assertions through Gradle

AllardQuek committed:
Fix minor styling issues

AllardQuek committed:
Add more Duke specific exceptions

There is only one Duke Exception to account for all types of Duke exceptions.

In order to change the exception message for one type of exception, edits would be required at multiple locations.

Adding new Classes to encapsulate common Duke Exceptions would allow us to more easily make changes to the exceptions being thrown.

AllardQuek committed:
Add common Duke exceptions

AllardQuek committed:
Ignore .vscode

AllardQuek committed:
Merge run configs

AllardQuek committed:
Integrate new exceptions

AllardQuek committed:
End files with new line

Files not ending with newline would cause checkstyle to complain.

Adding the newline for the newly created exception files would remove any errors due to the absence of a new line at the end of a file.

AllardQuek committed:
Convert TaskList to a static class

Calling TaskList methods in Parser requires getting the TaskList via Duke's getTasks() method, which is a bit unnecessary as the Parser does not need to know about Duke.

Making the TaskList fields and methods static allows us to simply call TaskList.method() and by pass having to get the TaskList via Duke.

AllardQuek committed:
Fix merge conflict

AllardQuek committed:
Add some assertions

AllardQuek committed:
Improve code quality

There are some cases where code quality was compromised due to prioritisation of the product's functionality.

Improving the code quality makes the code more readable and understandable, which includes:
* Updating java docs where necessary, especially for new exceptions
* using more intuitive variable names
* avoiding the use of magic numbers
* standardising implementation for similar methods
* using more appropriate access modifiers

AllardQuek committed:
Merge pull request #1 from AllardQuek/A-Assertions

Add some assertions

AllardQuek committed:
Fix merge conflict

AllardQuek committed:
Merge pull request #2 from AllardQuek/A-CodeQuality

A code quality

AllardQuek committed:
Create gradle.yml

AllardQuek committed:
Move .yml file

AllardQuek committed:
Fix some styles

AllardQuek committed:
Merge branch 'master' of https://github.com/AllardQuek/ip

AllardQuek committed:
Fix more styles

AllardQuek committed:
Attempt fix for checkstyle

AllardQuek committed:
Show all tasks in DialogBox

AllardQuek committed:
Remove redundant code and comment

AllardQuek committed:
Allow multiple deletes of tasks in one command

To delete multiple tasks, the user has to manually type the commands one by one for each task, which is repetitive and slow.

Supporting multiple deletes by allowing the user to key in space-separated task numbers to be deleted makes the user experience better.

This could also be easily extended to marking multiple tasks as done in a single command.

AllardQuek committed:
Merge pull request #3 from AllardQuek/C-MassOps

Type C feature: mass operations

AllardQuek committed:
Add unit tests to workflow

AllardQuek committed:
Set theme jekyll-theme-cayman

AllardQuek committed:
Add banner pic

AllardQuek committed:
Merge branch 'master' of https://github.com/AllardQuek/ip

AllardQuek committed:
Update README.md

AllardQuek committed:
Factor out saving of tasks to file

AllardQuek committed:
Add sample UI screenshot

AllardQuek committed:
Update README.md

AllardQuek committed:
Update README.md

AllardQuek committed:
Update README.md

AllardQuek committed:
Update README.md

AllardQuek committed:
Update README.md

AllardQuek committed:
Update README.md

AllardQuek committed:
Update README.md

AllardQuek committed:
Update README.md

AllardQuek committed:
Update README.md

AllardQuek committed:
Switch run from Text UI to GUI

AllardQuek committed:
Remove SingleSpaceSeparator

AllardQuek committed:
Minor reorg of code

AllardQuek committed:
Create new file if not exists

AllardQuek committed:
Return response instead of exception

AllardQuek committed:
Bug fixes and improve code quality

- Stricter checks for index out of bounds exception (when task number is 0 or negative)
- Update documentation
- Avoid magic numbers when deconstructing commands
- Add a couple of assertions

AllardQuek committed:
Refactor gui

AllardQuek committed:
Add documentation and clean up code

AllardQuek committed:
Remove unused import and fix styling issues

AllardQuek committed:
Add docs

AllardQuek committed:
Update README.md

AllardQuek committed:
Update README.md

AllardQuek committed:
Add final

AllardQuek committed:
Update README.md

AllardQuek committed:
Update README.md

AllardQuek committed:
Update README.md
